operation.go

```diff
@@ -35,12 +38,44 @@ func update(ctx context.Context, c *Collection, model Model, opts ...*options.Up
 		return err
 	}
 
-	res, err := c.UpdateOne(ctx, bson.M{field.ID: model.GetID()}, bson.M{"$set": model}, opts...)
+	query := bson.M{field.ID: model.GetID()}
+	modelVersionable, isVersionable := model.(Versionable)
+	var currentVersion interface{}
+	if isVersionable {
+		currentVersion = modelVersionable.GetVersion()
+
+		//Handle adding versionning for documents that were created without it
+		//In this case, the version field would be of zero value and not exist in the DB
+		//Add $or exists false in the query condition for this case
+		isCurrentVersionZero := false
+		switch c := currentVersion.(type) {
+		case string:
+			isCurrentVersionZero = c == ""
+		case int:
+			isCurrentVersionZero = c == 0
+		case time.Time:
+			isCurrentVersionZero = c.IsZero()
+		}
+
+		if isCurrentVersionZero {
+			query["$or"] = bson.A{bson.M{modelVersionable.GetVersionFieldName(): currentVersion}, bson.M{modelVersionable.GetVersionFieldName(): bson.M{"$exists": false}}}
+		} else {
+			query[modelVersionable.GetVersionFieldName()] = currentVersion
+		}
+
+		modelVersionable.IncrementVersion()
+	}
+
+	res, err := c.UpdateOne(ctx, query, bson.M{"$set": model}, opts...)
 
 	if err != nil {
 		return err
 	}
 
+	if isVersionable && res.MatchedCount == 0 {
+		return fmt.Errorf("versioning error : document %v %v with version %v could not be found", c.Name(), model.GetID(), currentVersion)
```

Comment:
Maybe it's good to convert it to a specific exported error  (e.g., `ErrVersionMismatch`), so users can check if the returned error is a version mismatch.

Reply:
Ok. I've done it in the following way so we can keep a useful message for the user while still being able to use errors.Is(err, ErrVersionning) : 
```
var ErrVersionning = errors.New("versionning error")

...
return fmt.Errorf("document %v %v with version %v could not be found %w", c.Name(), model.GetID(), v, ErrVersionning)
...
```